File: epi_judge_python/replace_and_remove.py
import functools
from typing import List

from test_framework import generic_test
from test_framework.test_utils import enable_executor_hook


def replace_and_remove(size: int, s: List[str]) -> int:
    na = nb = 0
    for i in range(size):
        if s[i] == 'a':
            na += 1
        elif s[i] == 'b':
            nb += 1
    count = size + na - nb
    idx = 0
    for i in range(size):
        if s[i] != 'b':
            s[idx] = s[i]
            idx += 1
    idx2 = count - 1
    for i in reversed(range(idx)):
        if s[i] == 'a':
            s[idx2] = s[idx2 - 1] = 'd'
            idx2 -= 2
        else:
            s[idx2] = s[i]
            idx2 -= 1
    return count


@enable_executor_hook
def replace_and_remove_wrapper(executor, size, s):
    res_size = executor.run(functools.partial(replace_and_remove, size, s))
    return s[:res_size]


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('replace_and_remove.py',
                                       'replace_and_remove.tsv',
                                       replace_and_remove_wrapper))

File: epi_judge_cpp/replace_and_remove.cc
#include <iterator>
#include <string>
#include <vector>

#include "test_framework/generic_test.h"
#include "test_framework/timed_executor.h"
using std::string;
using std::vector;

int ReplaceAndRemove(int size, char s[]) {
  int n = 0;
  for (auto i = 0; i < size; i++) {
    if (s[i] != 'b') {
      s[n++] = s[i];
    }
  }

  auto asize = 0;
  for (auto i = 0; i < n; i++) {
    asize += (s[i] == 'a') ? 2 : 1;
  }

  int idx = asize;
  for (auto i = n - 1; i >= 0; i--) {
    if (s[i] == 'a') {
      s[--idx] = 'd';
      s[--idx] = 'd';
    } else {
      s[--idx] = s[i];
    }
  }

  return asize;
}
vector<string> ReplaceAndRemoveWrapper(TimedExecutor &executor, int size,
                                       const vector<string> &s) {
  std::vector<char> s_copy(s.size(), '\0');
  for (int i = 0; i < s.size(); ++i) {
    if (!s[i].empty()) {
      s_copy[i] = s[i][0];
    }
  }

  int res_size =
      executor.Run([&] { return ReplaceAndRemove(size, s_copy.data()); });

  vector<string> result;
  for (int i = 0; i < res_size; ++i) {
    result.emplace_back(string(1, s_copy[i]));
  }
  return result;
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"executor", "size", "s"};
  return GenericTestMain(args, "replace_and_remove.cc",
                         "replace_and_remove.tsv", &ReplaceAndRemoveWrapper,
                         DefaultComparator{}, param_names);
}
